### PR TITLE
cloud-config: honor cloud_dir setting

### PR DESCRIPTION
**Pull request by aciba90**
## Proposed Commit Message

```
cloud-config: honor cloud_dir setting

- Modules affected: cmd.main, apport, devel.logs (collect-logs),
cc_snap, sources.DataSourceAzure, sources.DataSourceBigstep,
util:fetch_ssl_details.
- testing: Extend and port to pytest unit tests, add integration test.

LP: #1976564
```

## Additional Context

https://bugs.launchpad.net/cloud-init/+bug/1976564
https://warthogs.atlassian.net/browse/SC-1085

## Test Steps

```sh
cat > /tmp/95-custom-cloud-dir.cfg <<EOF
system_info:
  paths:
    cloud_dir: /new-cloud-dir
EOF

CONTAINER_NAME=dev-k
lxc launch ubuntu-daily:kinetic $CONTAINER_NAME
lxc file push /tmp/95-custom-cloud-dir.cfg $CONTAINER_NAME/etc/cloud/cloud.cfg.d/
lxc exec $CONTAINER_NAME -- rm -rf /var/lib/cloud
lxc exec $CONTAINER_NAME -- cloud-init clean --logs --reboot
lxc exec $CONTAINER_NAME -- cloud-init status --wait --long

# Expect no logs referencing use of /var/lib/cloud
lxc exec $CONTAINER_NAME -- grep /var/lib /var/log/cloud-init.log

# Expect logs referencing use of new cloud-dir
lxc exec $CONTAINER_NAME -- grep /new-cloud-dir var/log/cloud-init.log

# Expect to find no files in /var/lib/cloud
lxc exec $CONTAINER_NAME find /var/lib/cloud

# Expect user-data file pointing to the correct cloud-dir
lxc exec $CONTAINER_NAME -- cloud-init collect-logs -h
```

## Checklist:

 - [x] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/topics/contributing.html)
 - [x] I have updated or added any unit tests accordingly
 - [ ] I have updated or added any documentation accordingly

### CODE REVIEW

**Comment by TheRealFalcon** [comment]
@aciba90 , I noticed that this PR is in draft state. Is it still in progress, or is it ready for review?

**Comment by aciba90** [comment]
> @aciba90 , I noticed that this PR is in draft state. Is it still in progress, or is it ready for review?

Yes, it is, @TheRealFalcon. I want to run the new integration test on Azure. I will let you know when ready, thanks.

**Comment by aciba90** [comment]
I have tested the added integration test on Azure, and added the Azure mark to execute it there too. This PR is ready to be reviewed @TheRealFalcon, thanks.

File: tests/integration_tests/bugs/test_lp1976564.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    client.push_file(cfg, f"/etc/cloud/cloud.cfg.d/{CUSTOM_CLOUD_DIR_FN}")
+    client.execute(f"rm -rf {DEFAULT_CLOUD_DIR}")  # Remove previous cloud_dir
+    client.execute("cloud-init clean --logs")
+    client.restart()
+    yield client
+
+
+@pytest.mark.lxd_container
+@pytest.mark.lxd_vm
+@pytest.mark.azure
```
[the remark this answers is not in the document]
Comment (aciba90) [reply tests/integration_tests/bugs/test_lp1976564.py]: The only reason was to only cover the affected components in the bug. But it is better to unrestrict it to cover as much as possible, thanks for the suggestion.

File: tests/integration_tests/bugs/test_lp1976564.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+
+@pytest.fixture
+def custom_client(
+    client: IntegrationInstance, tmpdir
+) -> Iterator[IntegrationInstance]:
+    cfg = tmpdir.join(CUSTOM_CLOUD_DIR_FN)
+    with open(cfg, "w") as f:
+        f.write(CUSTOM_CLOUD_DIR)
+    client.push_file(cfg, f"/etc/cloud/cloud.cfg.d/{CUSTOM_CLOUD_DIR_FN}")
```
[the remark this answers is not in the document]
Comment (aciba90) [reply tests/integration_tests/bugs/test_lp1976564.py]: Thanks, much clearer with your suggestion!

File: cloudinit/sources/DataSourceBigstep.py [hunk 3]
```diff
@@ -31,22 +33,26 @@ def _get_data(self, apply_filter=False):
...
-        # If the file doesn't exist, then the server probably isn't a Bigstep
-        # instance; otherwise, another problem exists which needs investigation
-        if e.errno == errno.ENOENT:
-            return None
-        else:
-            raise
-    return content
+        return f"metadata ({self._get_url_from_file()})"
+
+    @lru_cache(maxsize=1, typed=False)
```
[the remark this answers is not in the document]
Comment (aciba90) [reply cloudinit/sources/DataSourceBigstep.py]: Done

File: tests/unittests/sources/test_bigstep.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
+import json
```
[the remark this answers is not in the document]
Comment (aciba90) [reply tests/unittests/sources/test_bigstep.py:1]: It was a pleasure.

**Comment by aciba90** [comment]
> Great work here! I left a few comments inline, but nothing major.

Thanks a lot, @TheRealFalcon! I have addressed your comments / suggestions and this PR is ready to be reviewed after the build passes.

**Review by TheRealFalcon — APPROVED** [review]
Thanks!

**Pull request merged**